### PR TITLE
Detect expired entries on read

### PR DESCRIPTION
**Pull request by 6543**
When reading sessions that should be expired expire them

Also fixes memory Init()

### CODE REVIEW

**Comment by 6543** [comment]
@unknwon the expire enhancements uported :)

File: file.go [hunk 1]
```diff
@@ -133,7 +133,15 @@ func (p *FileProvider) Read(sid string) (_ RawStore, err error) {
 	defer p.lock.RUnlock()
 
 	var f *os.File
+	ok := false
```
Comment (unknwon) [inline file.go:136]: I think rename this variable to be `expired` is better.

File: memory.go [hunk 2]
```diff
@@ -96,6 +96,8 @@ type MemProvider struct {
 // Init initializes memory session provider.
 func (p *MemProvider) Init(maxLifetime int64, _ string) error {
 	p.lock.Lock()
+	p.list = list.New()
+	p.data = make(map[string]*list.Element)
```
Comment (unknwon) [inline memory.go:100]: What is the fix/benefit here?

**Review by unknwon — CHANGES_REQUESTED** [review]
Thanks for the PR!

Can you add comments on every new addition about rationale?

File: memory.go (hunk 2 above)
Comment (zeripath) [reply memory.go:100]: This standardises the init function for the memprovider as compared to the other providers. 

All other providers the init function actually makes the provider ready - only the memprovider is different.

Say you wanted to have a doubly backed provider that wraps the eventual provider - for whatever reason - this glaring inconsistency breaks that.

File: file.go (hunk 1 above)
Comment (zeripath) [reply file.go:136]: That would require the following changes:

```suggestion
	expired := false
```

File: file.go [hunk 3]
```diff
@@ -133,7 +133,15 @@ func (p *FileProvider) Read(sid string) (_ RawStore, err error) {
...
 	if com.IsFile(filename) {
+		modTime, err := com.FileMTime(filename)
+		if err != nil {
+			return nil, err
+		}
+		ok = (modTime + p.maxlifetime) >= time.Now().Unix()
```
Comment (zeripath) [inline file.go:142]:
```suggestion
		expired = (modTime + p.maxlifetime) < time.Now().Unix()
```

File: file.go [hunk 4]
```diff
@@ -133,7 +133,15 @@ func (p *FileProvider) Read(sid string) (_ RawStore, err error) {
...
+	}
+	if ok {
```
Comment (zeripath) [inline file.go:144]:
```suggestion
	if !expired {
```

**Comment by 6543** [comment]
@unknwon does https://github.com/go-macaron/session/pull/41#discussion_r523398332 explain it :) ?

File: memory.go (hunk 2 above)
Comment (unknwon) [reply memory.go:100]: Gotcha, thanks!

File: memory.go [hunk 5]
```diff
@@ -130,7 +133,10 @@ func (p *MemProvider) Read(sid string) (_ RawStore, err error) {
 	// Create a new session.
 	p.lock.Lock()
 	defer p.lock.Unlock()
-
+	if ok {
+		p.list.Remove(e)
+		delete(p.data, sid)
```
Comment (unknwon) [inline memory.go:138]: This seems not necessary? `p.data[sid] = p.list.PushBack(s)` would override the value anyway.

**Comment by unknwon** [comment]
Thanks for the follow up! One last minor issue.

File: memory.go (hunk 5 above)
Comment (6543) [inline memory.go:138]:
```suggestion
```

**Comment by 6543** [comment]
@unknwon you are right, changed :)

**Review by unknwon — APPROVED** [review]
Thanks!

**Comment by unknwon** [comment]
https://github.com/go-macaron/session/releases/tag/v1.0.1 has been published for this merge.

**Pull request merged**